doc/jzip.6
.\" -*- nroff -*-
.TH JZIP 6 "" "2.1" "John\'s Z-code Interpreter Program"
.SH NAME
jzip \- execute Infocom v1-5 and Inform v1-5 & v8 game files
.SH SYNOPSIS
.B jzip
[
.I options
]
.I z-file-name
.SH DESCRIPTION
.B jzip
invokes the Z-code Interpreter.
.SH REQUIREMENTS
.B jzip
will run under most (if not all) flavors of Unix, including AIX, Solaris, 
SunOS, HP/UX, Linux, Free/NetBSD, IRIX, NeXT Mach, DEC OSF/1, and others.
.PP
.SH EDITING
.PP 
JZip now supports arrowkey editing in UNIX. 
.TP
.I Up/Down 
selects the previous / next command in history
.TP
.I Left/Right
moves the cursor over the line.
.TP
.I Backspace
deletes
.TP
.I Typing
inserts under the cursor
.SH OPTIONS
.TP
.I -l n
lines in display
.TP
.I -c n
columns in display
.TP
.I -r n
text right margin (default = 1)
.TP
.I -t n
text top margin (default = 0)
.TP
.I -v
display version information
.TP
.I -m
monochrome mode
.TP
.I -y
set Tandy bit
.TP
.I -s n
zmachine strict level, from 0 (none), 1 (report once, recommended) to 3.
.SH ENVIRONMENT
.B jzip
looks for the environment variable
.B INFOCOM_PATH
and will use the path found there to look for the story file specified on
the command line if the story file cannot be found in the current directory.
.SH CREDITS
.B jzip
is written and maintained by John Holder (j-holder@home.com), based on the
ZIP v2.0 sources by Mark Howell and Olaf 'Olsen' Barthel.
.SH AVAILABILITY
The most recent version of 
.B jzip
can be downloaded from one of:
.TP
http://www.jzip.sourceforge.net/
.TP
ftp://ftp.gmd.de/if-archive/infocom/interpreters/zip/
.SH FILES
.TP
.I /usr/games/jzip
The Z-Code Interpreter Program
.TP
.I /usr/lib/games/zfiles
Recommended location for Zcode files
.SH RELATED USENET NEWSGROUPS
.PD 0
Interactive Fiction (IF) authors should check out 
.I rec.arts.int-fiction,
and IF players should check out 
.I rec.games.int-fiction.




doc/History.txt

/* $Id: History.txt,v 1.2 2000/10/05 16:29:32 jholder Exp $ */

Jzip - Infocom Z-code Interpreter Program     V2.1
-----------------------------------------------------
Author: John Holder (j-holder@home.com)
Based on sources by Mark Howell & Olaf 'Olsen' Barthel

       This file contains a history of changes to the code base of Jzip.

Known Issues:
        - something is causing the comp2000 game "And the Waves Choke the Wind"
          by Gunther Schmidl to print text incorrectly. Finding and fixing
          this bug is top priority.
        - @aread/@sread is not quite up to spec.
        - adding support for Unicode opcodes/table per spec 1.0
          (currently, the hooks are in to look up the address of the table.
           nothing is done with it yet.)

Release 2.1, 04-Oct-2000
        - fixed bug in nested/redirected output streams when dealing w/ chars 
          outside the normal ASCII range (ie, ISO-Latin1).  This was causing
          the German translation of "Bear's Night Out" (bearg.z5) to display
          upside-down '?' instead of German eszett in some places.
        - renamed all opcode routines to z_opcodename to match functionality
          to Spec 1.0 more easily.

Release 2.1, 29-Jun-2000

        - NOW SPEC 1.0 COMPLIANT
        - fixed logical and arithmetic shift per spec 1.0 
        - fixed probelems with div and mod per spec 1.0 
        - rewrote much of property.c to handle 64 properties per spec 1.0 
        - fixed nested/redirected output streams per spec 1.0 (Xzip derived)
        - added support for extended load/save opcodes per spec 1.0 
        - added support for Quetzal portable save files (from Martin Frost)
               *** NOTE ***  Old save game files are now incompatible.
        - added ckifzs utility - checks if quetzal file is valid.
          improved to be more robust by John Kennedy.
        - added support for STRICTZ error checking (Xzip derived & extended)
        - @get_prop_len 0 returns zero per zmachine mailing list message
          from Graham.
        - Jzip's license is now very much less restrictive.
        - Added support for OS/2 and MS Quick C(DOS) (Thanks John Kennedy)
        - Can turn on Tandy bit (Thanks John Kennedy)
        - Can read Zcode files that have been compressed with gzip when
          built with -DUSE_ZLIB.  Requires the free Zlib, see
          http://www.cdrom.com/pub/infozip/zlib/

Release 2.0.1g, 5-Dec-1995:

        - fixed silly error I introduced in osdepend.c (the set_colours bug)
        - fixed extra include in fileio.c (string.h doesn't need double
          inclusion)
        - fixed "magic" declaration that caused problems with jzexe.c on
          RISC boxen.

Release 2.0.1g:

        - John Menichelli(menichel@pixi.com) has added text editing 
          features and a command history to bccio.c (the Borland
          DOS interface).  Thanks, John!
        - With some termcap hacking and John Menichelli's code, I added
          text editing and command history functionality to the UNIX
          ports of Jzip.
        - At the moment, the DOS and UNIX versions now have the correct 
          timed read key interval of 1/10th of a second, instead of the 
          1 second interval that used to be there.
        - Dancer (dancer@gateway1.brisnet.org.au) has added an Atari ST
          port of Jzip to the distribution.  He has also got the command
          history and editing features working.  If you have problems
          with this port, contact him.  Better documentation is in Jzip.txt
        - You now can create standalone DOS executable files with
          Magnus Olsson's utility JZexe, which he has kindly added to
          the Jzip distribution.

Release 2.0.1f:

        - Jzip now supports a default set of paths where it will look for
          a game file if it isn't in the current directory.  Set the
          environment variable INFOCOM_PATH to the directory or list of
          directories that contain your Zcode story files. For example:
          UNIX:
            setenv INFOCOM_PATH /usr/games/zfiles/:/usr/jholder/zfiles
          DOS:
            set INFOCOM_PATH=d:\infocom\ltoi1;d:\infocom\ltoi2;
          Suggested by Michael Phillips, (msphil@aardvark.cc.wm.edu).
        - Jzip has a manpage!  Written by Michael Phillips, I will edit
          and maintain it from now on.  Thanks, Mike!
        - UNIX versions no longer die a horrible death and dump core,
          cause bus errors, or put the screen in an incorrect mode when
          attempting to run an invalid file.
	  
Release 2.0.1e:

	- Jzip's UNIX distribution is now even more portable!  Some
	  problems with SunOS HP-UX, and Irix have been resolved, all
	  thanks to Mark Phillips (msp@bnr.co.uk).  As you port Jzip
	  to different versions of UNIX, please tell me what platform
	  and version of UNIX you are running, and send me a copy of
	  your makefile so I can document it. 
	- Jzip save files are now PORTABLE between big-endian and little-
	  endian machines!  Once again, thanks to Mark Phillips!  Now
	  you can play at home on your PC, upload your save file to work,
	  and play on your lunch hour! (Note: The 27-sept-1995 release
	  had a typo in the code & this didn't work! Sorry!)
	  *** NOTE ***  Old save game files are now incompatible on some boxes.

Release 2.0.1d:

       - Jzip properly understand what game you are playing and will
         "automagically" name your save games files, script files, and
         record files appropriately.  (ie. if you are playing zork1.dat,
         your save file is zork1.sav, script file is zork1.scr, and
         record file is zork1.rec)
       - The Borland DOS compile now supports color or monochrome
         mode with the -m command line option.  
       - Jzip now properly prints menus on the DOS compile.  Thanks
         for noticing the problem & sending in the fix to Andrew
         Benham. (A.D.S.Benham@bnr.co.uk)

Release 2.0.1a:


jzip.c

/* $Id: jzip.c,v 1.3 2000/10/04 23:07:57 jholder Exp $   
 * --------------------------------------------------------------------
 * see doc/License.txt for License Information   
 * --------------------------------------------------------------------
 * 
 * File name: $Id: jzip.c,v 1.3 2000/10/04 23:07:57 jholder Exp $  
 *   
 * Description:    
 *    
 * Modification history:      
 * $Log: jzip.c,v $
 * Revision 1.3  2000/10/04 23:07:57  jholder
 * fixed redirect problem with isolatin1 range chars
 *
 * Revision 1.2  2000/05/25 22:28:56  jholder
 * changes routine names to reflect zmachine opcode names per spec 1.0
 *
 * Revision 1.1.1.1  2000/05/10 14:21:34  jholder
 *
 * imported
 *
 *
 * --------------------------------------------------------------------
 */

/*
 * jzip.c
 *
 * Z code interpreter main routine. 
 *
 * Revisions list:
 * Mark Howell 10-Mar-1993 V2.0    howell_ma@movies.enet.dec.com
 *                         V2.0.1(some of a-f)
 * John Holder 22-Nov-1995 V2.0.1g j-holder@home.com
 * John Holder 06-Feb-1998 V2.0.2  j-holder@home.com Zstrict & Quetzal
 *
 * If you have problems with this interpreter and/or fix bugs in it, 
 * please notify John Holder (j-holder@home.com) and he will add your 
 * fix to the official JZIP distribution.
 */

#include "ztypes.h"
#include "jzexe.h"

extern int GLOBALVER;

static void configure( zbyte_t, zbyte_t );

/*
 * main
 *
 * Initialise environment, start interpreter, clean up.
 */

int main( int argc, char *argv[] )
{
   process_arguments( argc, argv );

   configure( V1, V8 );

   initialize_screen(  );

   load_cache(  );

   z_restart(  );

   ( void ) interpret(  );

   close_story(  );

   close_script(  );

   unload_cache(  );

   reset_screen(  );

   exit( EXIT_SUCCESS );

   return ( 0 );

}                               /* main */

/*
 * configure
 *
 * Initialise global and type specific variables.
 *
 */

static void configure( zbyte_t min_version, zbyte_t max_version )
{
   zbyte_t header[PAGE_SIZE], second;

   read_page( 0, header );
   datap = header;

   h_type = get_byte( H_TYPE );

   if ( h_type == 'M' || h_type == 'Z' )
   {
      /* possibly a DOS executable file, look more closely. (mol951115) */
      second = get_byte( H_TYPE + 1 );
      if ( ( h_type == 'M' && second == 'Z' ) || ( h_type == 'Z' && second == 'M' ) )
         if ( analyze_exefile(  ) )
         {
            /* Bingo!  File is a standalone game file.  */
            /* analyze_exefile has updated story_offset, so now we can */
            /* read the _real_ first page, and continue as if nothing  */
            /* has happened. */
            read_page( 0, header );
            h_type = get_byte( H_TYPE );
         }
   }

   GLOBALVER = h_type;

   if ( h_type < min_version || h_type > max_version ||
        ( get_byte( H_CONFIG ) & CONFIG_BYTE_SWAPPED ) )
      fatal( "Wrong game or version" );
   /*
    * if (h_type == V6 || h_type == V7)
    * fatal ("Unsupported zcode version.");
    */

   if ( h_type < V4 )
   {
      story_scaler = 2;
      story_shift = 1;
      property_mask = P3_MAX_PROPERTIES - 1;
      property_size_mask = 0xe0;
   }
   else if ( h_type < V8 )
   {
      story_scaler = 4;
      story_shift = 2;
      property_mask = P4_MAX_PROPERTIES - 1;
      property_size_mask = 0x3f;
   }
   else
   {
      story_scaler = 8;
      story_shift = 3;
      property_mask = P4_MAX_PROPERTIES - 1;
      property_size_mask = 0x3f;
   }

   h_config = get_byte( H_CONFIG );
   h_version = get_word( H_VERSION );
   h_data_size = get_word( H_DATA_SIZE );
   h_start_pc = get_word( H_START_PC );
   h_words_offset = get_word( H_WORDS_OFFSET );
   h_objects_offset = get_word( H_OBJECTS_OFFSET );
   h_globals_offset = get_word( H_GLOBALS_OFFSET );
   h_restart_size = get_word( H_RESTART_SIZE );
   h_flags = get_word( H_FLAGS );
   h_synonyms_offset = get_word( H_SYNONYMS_OFFSET );
   h_file_size = get_word( H_FILE_SIZE );
   if ( h_file_size == 0 )
      h_file_size = get_story_size(  );
   h_checksum = get_word( H_CHECKSUM );
   h_alternate_alphabet_offset = get_word( H_ALTERNATE_ALPHABET_OFFSET );

   datap = NULL;

}                               /* configure */
